devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java

```diff
@@ -0,0 +1,299 @@
+import io.quarkus.bootstrap.resolver.model.impl.WorkspaceModuleImpl;
+import io.quarkus.bootstrap.util.QuarkusModelHelper;
+import io.quarkus.gradle.tasks.QuarkusGradleUtils;
+import io.quarkus.runtime.LaunchMode;
+
+public class QuarkusModelBuilder implements ParameterizedToolingModelBuilder {
+
+    private static final List<String> scannedConfigurations = new LinkedList();
+
+    @Override
+    public boolean canBuild(String modelName) {
+        return modelName.equals(QuarkusModel.class.getName());
+    }
+
+    @Override
+    public Class getParameterType() {
+        return ModelParameter.class;
+    }
+
+    @Override
+    public Object buildAll(String modelName, Project project) {
+        final ModelParameterImpl modelParameter = new ModelParameterImpl();
+        modelParameter.setMode(LaunchMode.DEVELOPMENT.toString());
+        return buildAll(modelName, modelParameter, project);
+    }
+
+    @Override
+    public Object buildAll(String modelName, Object parameter, Project project) {
+        LaunchMode mode = LaunchMode.valueOf(((ModelParameter) parameter).getMode());
+
+        if (LaunchMode.TEST.equals(mode)) {
+            scannedConfigurations.add(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+        } else {
+            scannedConfigurations.add(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+        }
+
+        if (LaunchMode.DEVELOPMENT.equals(mode)) {
+            scannedConfigurations.add(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+        }
+
+        final Collection<org.gradle.api.artifacts.Dependency> directExtensionDependencies = getEnforcedPlatform(project);
+
+        final Map<ArtifactCoords, Dependency> appDependencies = new HashMap<>();
+        for (String configurationName : scannedConfigurations) {
+            final ResolvedConfiguration configuration = project.getConfigurations().getByName(configurationName)
+                    .getResolvedConfiguration();
+            appDependencies.putAll(collectDependencies(configuration, configurationName, mode, project));
+            directExtensionDependencies
+                    .addAll(getDirectExtensionDependencies(configuration.getFirstLevelModuleDependencies(), appDependencies));
+        }
+
+        final Set<Dependency> extensionDependencies = collectExtensionDependencies(project, directExtensionDependencies);
+
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+
+        return new QuarkusModelImpl(new WorkspaceImpl(appArtifactCoords, getWorkspace(project.getRootProject())),
+                new HashSet<>(appDependencies.values()),
+                extensionDependencies);
+    }
+
+    public Set<WorkspaceModule> getWorkspace(Project project) {
+        Set<WorkspaceModule> modules = new HashSet<>();
+        if (project.getSubprojects().isEmpty()) {
+            return Collections.singleton(getWorkspaceModule(project));
```

Comment:
The `project`'s module is not added to the `modules` otherwise.

Reply:
well, I think you right, as long as, it as some java plugin with source, it should be added